README
Information on this package -- including installation and usage instructions --
can be found in the package's documentation subdirectory, "html/" or on-line at
<http://www.unidata.ucar.edu/software/noaaport>.


CHANGE_LOG
1.7.1 	2011-06-16 16:45:37-06:00
    Added "Preinstallation" section to the documentation. Contains commentary
    on changing the kernel parameter "net.ipv4.ipfrag_max_dist" from its
    default value on Redhat systems.

    Improved installation instructions:
        Added commentary about stopping the LDM system.
        Added "noaaportIngester" to rationale for giving root password.

    Improved the automatic setting of the "prefix" variable by the configure(1)
    script.

1.7.0 	2011-06-06 10:51:36-06:00
    noaaportIngester(1):
        Created this program, which is a merger of dvbs_multicast(1) and
        readnoaaport(1) into one, multi-threaded program.

    Misc:
        Deleted hdlcrecv(1)-using scripts.

        Converted scripts to use noaaportIngester(1).

        Moved getFaciltyName() to its own file in order to port to Solaris.

        Made reporting of products a NOTE level message.

        Added 2 GEMPAK tables.

1.6.2 	2011-05-04 12:28:49-06:00
    Improved release-engineering rules in makefile.

1.6.1	2011-05-03 15:51:08-06:00
    Corrected use of package and version strings in top-level webpage.

    Improved release procedure.

    Improved portability of build procedure:

1.6.0	2011-05-03 10:35:12-06:00
    Installation:
        Migrated to automake(1)-based installation.

    dvbs_multicast(1):
        Modified to use blocking shmfifo_put().

        Improved information hiding.

        Improved logging:
            Added size of gap to gap log messages.

            Added logging of retrograde packet numbers.
        
            Modified to log correctly while processing command-line arguments.

        Improved portability:
            Added use of HAVE_GET_QUEUE_PATH so that it can build against
            LDM packages prior to 6.9.

        Reformatted to improve clarity.

    readnoaaport(1):
        Modified to use blocking shmfifo_get().

        Rewrote _shm_bufread() to improve clarity.

        Corrected initialization of "struct shmhandle*".

        Improved information hiding.

        Modified to log correctly while processing command-line arguments.

        Improved portability:
            Added "#define __EXTENSIONS__" so that SunOS 5.10's
            /opt/SUNWspro/bin/cc(1) will compile it.

            Added function prototype for "close_pq()".

            Added use of HAVE_GET_QUEUE_PATH so that it can build against
            LDM packages prior to 6.9.

        Reformatted to improve clarity.

        Added some internal documentation.

    shmfifo(3) module:
        Corrected assignment bug.

        Reworked semaphore signaling. Before, only 1 semaphore was used.
        Now, 3 semaphores are used: one for locking, one as a writer
        condition-variable, and one as a reader condition-variable.

        Improved shmfifo_put():
            Made blocking.

            Made writing to the FIFO event-driven rather than poll-driven.

            Added test for invalid size argument and for messages that are
            too big to put into the FIFO.

            Ensured unconditional unlocking of FIFO.

            Added once-per-call logging of insufficient FIFO space.
            
            Reformatted to structured programing.

            Added internal documentation.

        Improved shmfifo_get():
            Made blocking.

            Made reading from the FIFO event-driven rather than poll-driven.

            Ensured unconditional unlocking of FIFO.
            
            Reformatted to structured programing.

            Added internal documentation.

        Added shmfifo_wait(), shmfifo_new(), shmfifo_free(), checkLocked() and
        checkUnlocked().

        Reformatted shmfifo_create() to improve clarity.

        Replaced "shmfifo.h" with "shmfifo.hin" and extractDecls(1).

        Added "const" to many places.

    Misc:
        Added README file.


readpsh.c
/*
 *   Copyright 2004, University Corporation for Atmospheric Research
 *   See COPYRIGHT file for copying and redistribution conditions.
 */
#include <stdio.h>
#include <string.h>
#include <ulog.h>
#include "nport.h"

int wmo_header(char *prod_name, size_t prod_size, char *wmohead, char *wmometa, int *metaoff);
void redbook_header(char *buf, int nbytes, char *tstr);

int readpsh(char *buf,psh_struct *psh)
{
int iret=0;
unsigned char b1,b2,b3,b4,val;
unsigned long lval;

psh->hasccb = 0;
psh->ccbmode = 0;
psh->ccbsubmode = 0;
psh->metadata[0] = '\0';
psh->metaoff = -1;

b1 = (unsigned char)buf[0];
b2 = (unsigned char)buf[1];
b3 = (unsigned char)buf[2];
b4 = (unsigned char)buf[3];
lval = (b3 << 8) + b4;   /*psh len */

psh->onum = b1;
psh->otype = b2;
psh->olen = lval;

val = (unsigned char)buf[4];
psh->version = val;

val = (unsigned char)buf[5];
psh->hflag = val;

b1 = (unsigned char)buf[6];
b2 = (unsigned char)buf[7];
lval = (b1 << 8) + b2;
psh->psdl = lval;

b1 = (unsigned char)buf[8];
b2 = (unsigned char)buf[9];
lval = (b1 << 8) + b2;
psh->bytes_per_record = lval;

val = (unsigned char)buf[10];
psh->ptype = val;

val = (unsigned char)buf[11];
psh->pcat = val;

b1 = (unsigned char)buf[12];
b2 = (unsigned char)buf[13];
lval = (b1 << 8) + b2;
psh->pcode = lval;

b1 = (unsigned char)buf[14];
b2 = (unsigned char)buf[15];
lval = (b1 << 8) + b2; 
psh->frags = lval;

b1 = (unsigned char)buf[16];
b2 = (unsigned char)buf[17];
lval = (b1 << 8) + b2;
psh->nhoff = lval;

b1 = (unsigned char)buf[19];
psh->source = b1;

b1 = (unsigned char)buf[20];
b2 = (unsigned char)buf[21];
b3 = (unsigned char)buf[22];
b4 = (unsigned char)buf[23];
lval = (((((b1 << 8) + b2) << 8) + b3) << 8) + b4;
psh->seqno = lval;

b1 = (unsigned char)buf[24];
b2 = (unsigned char)buf[25];
b3 = (unsigned char)buf[26];
b4 = (unsigned char)buf[27];
lval = (((((b1 << 8) + b2) << 8) + b3) << 8) + b4;
psh->rectime = lval;

b1 = (unsigned char)buf[28];
b2 = (unsigned char)buf[29];
b3 = (unsigned char)buf[30];
b4 = (unsigned char)buf[31];
lval = (((((b1 << 8) + b2) << 8) + b3) << 8) + b4;
psh->transtime = lval;

b1 = (unsigned char)buf[32];
b2 = (unsigned char)buf[33];
lval = (b1 << 8) + b2;
psh->runid = lval;

b1 = (unsigned char)buf[34];
b2 = (unsigned char)buf[35];
lval = (b1 << 8) + b2;
psh->origrunid = lval;

return(iret);
}

int readccb(char *buf,ccb_struct *ccb, psh_struct *psh, int blen)
{
int iret=0;
unsigned char b1,b2;
char wmohead[256];
int wmolen, metaoff=-1;
char redbook_title[45];
char wmometa[256];

redbook_title[0] = '\0';
wmometa[0] = '\0';

b1 = (unsigned char)buf[0];
b2 = (unsigned char)buf[1];

ccb->b1 = b1;
ccb->len = 2 * (int)(((b1 & 63) << 8) + b2);

if ( ccb->len > blen ) /* A rouge product missing its CCB, reported to NWS DM 3/10/05 by Chiz */
   {
   uerror ("invalid ccb length = %d %d %d, blen %d\n",ccb->len,b1,b2,blen);

   /* try a failsafe header, otherwise use our own! */
   wmolen = 0;
   while ( ( (int)buf[wmolen] >= 32) && ( wmolen < 256 ) ) wmolen++;
   if ( wmolen > 0 )
      strncat(psh->pname,buf,wmolen);
   else
      sprintf(psh->pname,"Unidentifiable product\0");
   ccb->len = 0;
   return(-1);
   }


b1 = (unsigned char)buf[10];
b2 = (unsigned char)buf[11];
psh->ccbmode = b1;
psh->ccbsubmode = b2;
udebug("ccb mode %d ccb submode %d\0",psh->ccbmode,psh->ccbsubmode);
psh->hasccb = 1;

b1 = (unsigned char)buf[12];
b2 = (unsigned char)buf[13];
ccb->user1 = b1;
ccb->user2 = b1;
udebug("ccb user1 %d ccb user2 %d\0",ccb->user1,ccb->user2);
psh->ccbdtype[0] = '\0';     /* Initialize ccbdtype...eventually used to identify data type */

/* see if this looks like a WMO header, if so canonicalize */
if(wmo_header(buf+ccb->len,blen - ccb->len,wmohead,wmometa,&metaoff)==0)
   {
   if(strlen(wmohead) > 0)
      {
      strcat(psh->pname,wmohead); 
      if (metaoff > 0) psh->metaoff = metaoff;
      }
   else
      {
      wmolen = 0;
      while ( ( (int)buf[wmolen + ccb->len] >= 32) && ( wmolen < 256 ) )
         wmolen++;

      strncat(psh->pname,buf+ccb->len,wmolen);
      }
   }
else
   {
   wmolen = 0;
   while ( ( (int)buf[wmolen + ccb->len] >= 32) && ( wmolen < 256 ) )
      wmolen++;
   if ( wmolen > 0 )
      strncat(psh->pname,buf+ccb->len,wmolen);
   else
      sprintf(psh->pname,"Unidentifiable product\0");

   unotice("Non-wmo product type %s ccbmode %d ccbsubmode %d\0",
	psh->pname, psh->ccbmode, psh->ccbsubmode);
   }


/* for now, since NEXRAD is not identified as a separate PTYPE, use WMO header check */
if((psh->ccbmode == 2)&&(psh->ccbsubmode == 0)) 
   {
   if((memcmp(buf+ccb->len,"SDUS5",5) == 0)||
      (memcmp(buf+ccb->len,"SDUS2",5) == 0)||
      (memcmp(buf+ccb->len,"SDUS3",5) == 0)||
      (memcmp(buf+ccb->len,"SDUS7",5) == 0)||
      (memcmp(buf+ccb->len,"SDUS8",5) == 0) )
      {
      /* uncompressed nids check */
      if(ccb->user1 != 'F') sprintf(psh->ccbdtype,"nids/\0");
      psh->pcat = 99;
      }
   }
/*
unfortunately, NWS isn't using CCB mode for all product types
else if((psh->pcat == 4)&&(psh->ccbmode == 2)&&(psh->ccbsubmode == 1))
   sprintf(psh->ccbdtype,"grib/\0");
else if(psh->pcat == 5)
   sprintf(psh->ccbdtype,"bufr/\0");
*/
else if (psh->pcat == 2)
   {
   /* see if this is recognizeable as a redbook graphic */
   redbook_header(buf+ccb->len,blen - ccb->len, redbook_title);
   if(strlen(redbook_title) > 0)
      sprintf(psh->ccbdtype,"redbook %d_%d/\0", ccb->user1, ccb->user2);
   else
      sprintf(psh->ccbdtype,"graph %d_%d/\0",  ccb->user1, ccb->user2);
   }



/* create metadata */
sprintf(psh->metadata," !\0");

if ( psh->ccbdtype[0] != '\0')
   strcat ( psh->metadata, psh->ccbdtype);

if ( redbook_title[0] != '\0' )
   strcat( psh->metadata, redbook_title);

if ( wmometa[0] != '\0' )
   strcat( psh->metadata, wmometa);

if ( psh->metadata[2] == '\0' ) psh->metadata[0] = '\0';


/* restore meta data to PROD NAME
strcat(psh->pname,psh->metadata); */

return(iret);
}
